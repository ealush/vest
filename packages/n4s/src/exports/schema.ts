import { DropFirst } from 'utilityTypes';

import { isArrayOf } from 'isArrayOf';
import { loose } from 'loose';
import { enforce } from 'n4s';
import { optional } from 'optional';
import { shape } from 'shape';

export { partial } from 'partial';

enforce.extend({ isArrayOf, loose, optional, shape });

/* eslint-disable @typescript-eslint/no-namespace */
declare global {
  namespace n4s {
    interface EnforceCustomMatchers<R> {
      isArrayOf: (...args: DropFirst<Parameters<typeof isArrayOf>>) => R;
      loose: (...args: DropFirst<Parameters<typeof loose>>) => R;
      shape: (...args: DropFirst<Parameters<typeof shape>>) => R;
      optional: (...args: DropFirst<Parameters<typeof optional>>) => R;
    }
  }
}
