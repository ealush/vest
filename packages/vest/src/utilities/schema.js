import createContext from 'context';

import isFunction from 'isFunction';
import throwError from 'throwError';

// Matches either a dot (.) inside a string
// Or a number wrapped in brackets for simple
// path lookup
const toPathRegexp = /\.|\[([\d)]*)\]/g;

const context = createContext(() => ({
  include: [],
  exclude: [],
}));

/**
 * A wrapper function for an enforce schema. It gives
 * back an interface that's similar to Vest's result object
 * that will make it easier to query the validation results
 *
 * @param {Object} enforceSchema
 * @param {Function} [body]
 */
function schema(enforceSchema, body) {
  return context.bind({}, function (...args) {
    if (isFunction(body)) {
      body(...args);
    }

    if (!isFunction(enforceSchema.run)) {
      throwError('Not a valid enforce schema.');
    }
    const result = enforceSchema.run(...args);

    const parsedResult = transformResultObject(result);

    let _hasErrors, _hasWarnings;

    const errorMessages = {};
    const warningMessages = {};
    const tests = {};

    const { include, exclude } = context.use();

    for (const test in parsedResult) {
      const isExcluded = exclude.some(name => test.startsWith(name));

      const isIncluded = include.some(name => test.startsWith(name));

      if (isExcluded || (include.length && !isIncluded)) {
        continue;
      }

      tests[test] = parsedResult[test];

      errorMessages[test] = parsedResult[test].errors = Object.keys(
        parsedResult[test].errors
      );
      warningMessages[test] = parsedResult[test].warnings = Object.keys(
        parsedResult[test].warnings
      );
      _hasErrors = _hasErrors || parsedResult[test].hasErrors;
      _hasWarnings = _hasWarnings || parsedResult[test].hasWarnings;
    }

    const getFn = (key, storage) =>
      !key ? storage : getKeyFromResults(key, storage) || [];

    function getErrors(key) {
      return getFn(key, errorMessages);
    }

    function getWarnings(key) {
      return getFn(key, warningMessages);
    }

    function hasErrors(key) {
      return key ? !!errorMessages[key]?.length : _hasErrors;
    }

    function hasWarnings(key) {
      return key ? !!warningMessages[key]?.length : _hasWarnings;
    }

    return {
      getErrors,
      getWarnings,
      hasErrors,
      hasWarnings,
      tests,
    };
  });
}

function getKeyFromResults(key, storage) {
  // If the provided key exists in the object, return as is.
  if (storage[key]) {
    return storage[key];
  }

  const parsed = key.split(toPathRegexp).reduce((path, current) => {
    // If current path segment is falsy, continue
    if (!current) {
      return path;
    }

    const parsedInt = parseInt(current);

    // If the path segment equals (==) its parseInt result, it is a number
    const isNumber = parsedInt == current;

    // If the path is still empty
    if (!path) {
      // return the current segment
      return path + (isNumber ? parsedInt : current);
    }

    // if number, wrap with a brackets, else, concatenate with a period
    return path + (isNumber ? `[${parsedInt}]` : '.' + current);
  }, '');

  // return parsed path
  return storage[parsed];
}

// Parses the result object
function transformResultObject(root) {
  const tests = {};
  function traverse(node, key = '') {
    if (key) {
      tests[key] = {
        errors: [],
        hasErrors: node.hasErrors,
        hasWarnings: node.hasWarnings,
        warnings: [],
      };
    }

    // Extract the error and warning messages from each node
    if (node.failed && node.message) {
      if (node.warn) {
        node.warnings = Object.assign({}, node.warnings, {
          [node.message]: node.message,
        });
      } else {
        node.errors = Object.assign({}, node.errors, {
          [node.message]: node.message,
        });
      }
    }

    // traverse each child
    for (const childKey in node.children) {
      const child = node.children[childKey];

      // Traverse each child node
      traverse(
        child,
        !key
          ? childKey
          : [
              key,
              node.isArray ? `[${childKey}]` : `${key ? '.' : ''}${childKey}`,
            ].join('')
      );

      // concatenate the node's messages with its childrens' messages
      node.warnings = Object.assign({}, node.warnings, child.warnings);
      node.errors = Object.assign({}, node.errors, child.errors);
    }

    // Update the shared tests object
    if (key) {
      tests[key].warnings = Object.assign(
        {},
        tests[key].warnings,
        node.warnings
      );
      tests[key].errors = Object.assign({}, tests[key].errors, node.errors);
    }
  }

  traverse(root);

  return tests;
}

export default Object.defineProperties(schema, {
  skip: {
    value: function skip(namespace) {
      const { exclude } = context.use();
      exclude.push(namespace);
    },
  },
  only: {
    value: function only(namespace) {
      const { include } = context.use();
      include.push(namespace);
    },
  },
});
