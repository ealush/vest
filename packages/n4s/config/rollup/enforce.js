const genConfig = require('../../../../config/rollup/genConfig');
const { packageDist, packageNames } = require('../../../../util');

export default genConfig({
  libraryName: packageNames.N4S,
  distPath: packageDist(packageNames.N4S),
  input: 'enforce.js',
});
