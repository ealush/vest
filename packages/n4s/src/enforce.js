import bindLazyRule from 'bindLazyRule';
import bindExtend from 'enforce.extend';
import bindTemplate from 'enforce.template';
import runner from 'enforceRunner';
import genRuleProxy from 'genRuleProxy';
import runtimeRules from 'runtimeRules';
import withArgs from 'withArgs';

const Enforce = value => {
  const proxy = genRuleProxy({}, ruleName =>
    withArgs(args => {
      runner(runtimeRules[ruleName], value, args);
      return proxy;
    })
  );
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

bindExtend(enforce, Enforce);
bindTemplate(enforce);

export default enforce;
