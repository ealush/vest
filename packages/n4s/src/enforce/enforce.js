import compounds from 'compounds';
import { RUN_RULE } from 'enforceKeywords';
import runner from 'enforceRunner';
import isFunction from 'isFunction';
import isRule from 'isRule';
import proxySupported from 'proxySupported';
import rules from 'rules';

const rulesObject = Object.assign(rules(), compounds);

let rulesList = proxySupported() ? null : Object.keys(rulesObject);

const Enforce = value => {
  const proxy = genRuleProxy(enforce, ruleName => (...args) => {
    runner(rulesObject[ruleName], value, args);
    return proxy;
  });
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
    genRuleProxy(Enforce, bindLazyRule);
  }

  return enforce;
};

export default enforce;

// Creates a proxy object that has access to all the rules
function genRuleProxy(target, output) {
  if (proxySupported()) {
    return new Proxy(target, {
      get: (target, fnName) => {
        if (isRule(rulesObject, fnName)) {
          return output(fnName);
        }

        return target[fnName];
      },
    });
  } else {
    /**
     * This method is REALLY not recommended as it is slow and iterates over
     * all the rules for each direct enforce reference. We only use it as a
     * lightweight alternative for the much faster proxy interface
     */
    return rulesList.reduce((target, fnName) => {
      return Object.defineProperties(target, {
        [fnName]: { get: () => output(fnName), configurable: true },
      });
    }, target);
  }
}

// Initiates a chain of functions directly from the `enforce`
// function - that's even though we do not have any closure
// there to store that data.
function bindLazyRule(ruleName) {
  const registeredRules = [];

  const addFn = fnName => (...args) => {
    registeredRules.push(
      Object.defineProperty(
        value => rulesObject[fnName](value, ...args),
        'name',
        { value: fnName }
      )
    );

    const returnvalue = genRuleProxy({}, addFn);

    return Object.assign(returnvalue, {
      [RUN_RULE]: getValue => {
        return registeredRules.every(fn => {
          // This  inversion of control when getting the value is
          // required in order to pass the function over to `shape`
          // so it can make the decision which args to pass to `optional`
          return fn(isFunction(getValue) ? getValue(fn.name) : getValue);
        });
      },
    });
  };

  return addFn(ruleName);
}
