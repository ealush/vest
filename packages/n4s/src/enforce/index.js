import { isRule, proxySupported } from '../lib';
import rules from '../rules';
import runner from './runner';

const rulesObject = { ...rules };

let enforce, rulesList;

if (proxySupported()) {
  enforce = value => {
    const proxy = new Proxy(rulesObject, {
      get: (rules, fnName) => {
        if (!isRule(rules, fnName)) {
          return;
        }

        return (...args) => {
          runner(rules[fnName], value, ...args);
          return proxy;
        };
      },
    });
    return proxy;
  };
} else {
  rulesList = Object.keys(rulesObject);

  enforce = value =>
    rulesList.reduce(
      (allRules, fnName) =>
        Object.assign(allRules, {
          ...(isRule(rulesObject, fnName) && {
            [fnName]: (...args) => {
              runner(rulesObject[fnName], value, ...args);
              return allRules;
            },
          }),
        }),
      {}
    );
}

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
  }
};

export default enforce;
