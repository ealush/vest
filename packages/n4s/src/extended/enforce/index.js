import enforce from '../../enforce';
import rules, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
} from '../rules';

enforce.extend(rules);

export default Object.assign(enforce, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
});
