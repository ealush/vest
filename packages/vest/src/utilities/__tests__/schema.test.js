import enforce from 'enforce';
import schema from 'schema';

describe('schema', () => {
  it('Should correctly present test results', () => {
    const Schema = schema(User);
    const res = Schema({
      user: {
        name: { first: {}, last: 'last' },
      },
      id: 'x.',
      friends: [
        {
          name: {
            first: 'first_name',
            last: 1,
            middle: [],
            username: 0,
          },
        },
      ],
    });

    expect(res.hasErrors()).toBe(true);
    expect(res.hasWarnings()).toBe(true);
    expect(res.hasErrors('user')).toBe(true);
    expect(res.hasWarnings('user')).toBe(false);
    expect(res.hasErrors('user.name')).toBe(true);
    expect(res.hasWarnings('user.name')).toBe(false);
    expect(res.hasErrors('user.name.first')).toBe(true);
    expect(res.hasErrors('user.name.last')).toBe(false);
    expect(res.hasErrors('user.name.middle')).toBe(false);
    expect(res.hasErrors('id')).toBe(true);
    expect(res.hasErrors('friends')).toBe(true);
    expect(res.hasErrors('friends[0]')).toBe(true);
    expect(res.hasErrors('friends[0].name')).toBe(true);
    expect(res.hasErrors('friends[0].name.first')).toBe(false);
    expect(res.hasErrors('friends[0].name.last')).toBe(false);

    expect(res).toMatchSnapshot();
  });

  it('Should bubble error and warning messages up to the ancestor', () => {
    const Schema = schema(User);
    const res = Schema({
      user: {
        name: { first: null, last: null },
      },
      id: null,
      friends: [
        {
          name: {
            first: null,
            last: null,
            middle: null,
          },
        },
        {
          name: {
            first: null,
            last: null,
            middle: null,
          },
        },
      ],
    });

    expect(res.getErrors('user.name')).toEqual([
      ...res.getErrors('user.name.username'),
      ...res.getErrors('user.name.first'),
      ...res.getErrors('user.name.last'),
      ...res.getErrors('user.name.middle'),
    ]);
    expect(res.getWarnings('user.name')).toEqual(
      res.getWarnings('user.name.last')
    );
    expect(res.getErrors('friends')).toEqual([
      ...res.getErrors('friends[0].id'),
      ...res.getErrors('friends[0].name.first'),
      ...res.getErrors('friends[0].name.last'),
      ...res.getErrors('friends[0].name.middle'),
    ]);
  });

  it('Should return error messages when they should appear', () => {
    const Name = enforce.loose({
      name: enforce
        .loose({
          first: enforce.isString().isNotEmpty().message('first_name_error'),
          last: enforce.isString().isNotEmpty().message('last_name_error'),
        })
        .message('name_error'),
    });

    const Schema = schema(Name);
    const res = Schema({ name: {} });

    expect(res.hasErrors()).toBe(true);
    expect(res.getErrors()).toEqual({
      name: ['name_error', 'first_name_error', 'last_name_error'],
      'name.first': ['first_name_error'],
      'name.last': ['last_name_error'],
    });
  });
});

describe('schema function body', () => {
  const Shape = enforce.shape({
    a: enforce.isNumber().message('"a" must be a number'),
    b: enforce.isString().isNumeric().message('"b" must be a numeric string'),
    c: enforce.isArrayOf(
      enforce.shape({ id: enforce.isString() }).message('id must be a string')
    ),
  });

  const data = {
    a: 'not_a_number',
    b: 22,
    c: [{ id: 1 }, { id: 2 }, { id: '3' }],
  };

  it('passes all arguments to function body', () => {
    const body = jest.fn();

    const Schema = schema(Shape, body);
    Schema(data, 'args1', 'args2', 'etc');

    expect(body).toHaveBeenCalledWith(data, 'args1', 'args2', 'etc');
  });

  describe('schema.skip', () => {
    it('Should validate all but skipped fields', () => {
      const data = {
        a: 'not_a_number',
        b: 22,
        c: [{ id: 1 }, { id: 2 }, { id: '3' }],
      };
      {
        const Schema = schema(Shape, () => {
          schema.skip('a');
        });
        const result = Schema(data);
        expect(result.getErrors('a')).toEqual([]);
        expect(result.hasErrors('a')).toBe(false);
        expect(result.getErrors('b')).toEqual(['"b" must be a numeric string']);
        expect(result.hasErrors('b')).toBe(true);
        expect(result.tests).not.toHaveProperty('a');
        expect(result.tests.b).toBeDefined();
        expect(result.tests.c).toBeDefined();
        expect(result.tests['c[0]']).toBeDefined();
        expect(result.tests['c[0].id']).toBeDefined();
        expect(result.tests['c[1].id']).toBeDefined();
        expect(result.tests['c[2].id']).toBeDefined();
        expect(result.tests).toMatchSnapshot();
        expect(result.getErrors()).toMatchSnapshot();
      }
      {
        const Schema = schema(Shape, () => {
          schema.skip('c');
        });
        const result = Schema(data);
        expect(result.getErrors('a')).toEqual(['"a" must be a number']);
        expect(result.hasErrors('a')).toBe(true);
        expect(result.hasErrors('c')).toBe(false);
        expect(result.getErrors('c')).toEqual([]);
        expect(result.hasErrors('c[0]')).toBe(false);
        expect(result.hasErrors('c[0].id')).toBe(false);
        expect(result.tests.a).toBeDefined();
        expect(result.tests.b).toBeDefined();
        expect(result.tests).not.toHaveProperty('c');
        expect(result.tests).not.toHaveProperty('c[0]');
        expect(result.tests).not.toHaveProperty('c[1]');
        expect(result.tests).not.toHaveProperty('c[2]');
        expect(result.tests).not.toHaveProperty('c[0].id');
        expect(result.tests).not.toHaveProperty('c[1].id');
        expect(result.tests).not.toHaveProperty('c[2].id');
        expect(result.tests).toMatchSnapshot();
        expect(result.getErrors()).toMatchSnapshot();
      }
      {
        const Schema = schema(Shape, () => {
          schema.skip('a');
          schema.skip('b');
        });
        const result = Schema(data);
        expect(result.tests.a).not.toBeDefined();
        expect(result.tests.b).not.toBeDefined();
        expect(result.tests.c).toBeDefined();
        expect(result.tests).toMatchSnapshot();
        expect(result.getErrors()).toMatchSnapshot();
      }
    });
  });
  describe('schema.only', () => {
    it('Should only validate included fields', () => {
      const data = {
        a: 'not_a_number',
        b: 22,
        c: [{ id: 1 }, { id: 2 }, { id: '3' }],
      };
      {
        const Schema = schema(Shape, () => {
          schema.only('a');
        });
        const result = Schema(data);
        expect(result.hasErrors('a')).toBe(true);
        expect(result.getErrors('a')).toEqual(['"a" must be a number']);
        expect(result.hasErrors('b')).toBe(false);
        expect(result.hasErrors('c')).toBe(false);
        expect(result.tests.a).toBeDefined();
        expect(result.tests.b).not.toBeDefined();
        expect(result.tests.c).not.toBeDefined();
        expect(result.tests['c[0]']).not.toBeDefined();
        expect(result.tests['c[0].id']).not.toBeDefined();
        expect(result.tests['c[1].id']).not.toBeDefined();
        expect(result.tests['c[2].id']).not.toBeDefined();
        expect(result.tests).toMatchSnapshot();
        expect(result.getErrors()).toMatchSnapshot();
      }
    });
  });
});

const Name = enforce.loose({
  first: enforce.isString().message('First name must be a string'),
  last: enforce.isString().warn().message('last name must be a string'),
  middle: enforce
    .optional(enforce.isString().message('middle must be a string'))
    .message('if present, middle name must be a string'),
});

const Entity = enforce.loose({
  id: enforce
    .isString()
    .startsWith('x0.')
    .message('id must be a string starting with x0.'),
});

const Friends = enforce.isArrayOf(
  enforce.template(
    enforce.loose({
      name: Name,
    }),
    Entity
  )
);
const User = enforce.template(
  enforce.loose({
    user: enforce.shape({
      name: enforce.template(
        Name,
        enforce.loose({
          username: enforce.isString().message('User name must be a string'),
        })
      ),
    }),
    friends: Friends,
  }),
  Entity
);
