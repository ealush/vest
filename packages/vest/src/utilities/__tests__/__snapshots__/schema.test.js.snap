// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema Should correctly present test results 1`] = `
Object {
  "getErrors": [Function],
  "getWarnings": [Function],
  "hasErrors": [Function],
  "hasWarnings": [Function],
  "tests": Object {
    "friends": Object {
      "errors": Array [
        "id must be a string starting with x0.",
        "middle must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": true,
      "warnings": Array [
        "last name must be a string",
      ],
    },
    "friends[0]": Object {
      "errors": Array [
        "id must be a string starting with x0.",
        "middle must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": true,
      "warnings": Array [
        "last name must be a string",
      ],
    },
    "friends[0].id": Object {
      "errors": Array [
        "id must be a string starting with x0.",
      ],
      "hasErrors": true,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "friends[0].name": Object {
      "errors": Array [
        "middle must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": true,
      "warnings": Array [
        "last name must be a string",
      ],
    },
    "friends[0].name.first": Object {
      "errors": Array [],
      "hasErrors": false,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "friends[0].name.last": Object {
      "errors": Array [],
      "hasErrors": false,
      "hasWarnings": true,
      "warnings": Array [
        "last name must be a string",
      ],
    },
    "friends[0].name.middle": Object {
      "errors": Array [
        "middle must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "id": Object {
      "errors": Array [
        "id must be a string starting with x0.",
      ],
      "hasErrors": true,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "user": Object {
      "errors": Array [
        "User name must be a string",
        "First name must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "user.name": Object {
      "errors": Array [
        "User name must be a string",
        "First name must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "user.name.first": Object {
      "errors": Array [
        "First name must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "user.name.last": Object {
      "errors": Array [],
      "hasErrors": false,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "user.name.middle": Object {
      "errors": Array [],
      "hasErrors": false,
      "hasWarnings": false,
      "warnings": Array [],
    },
    "user.name.username": Object {
      "errors": Array [
        "User name must be a string",
      ],
      "hasErrors": true,
      "hasWarnings": false,
      "warnings": Array [],
    },
  },
}
`;

exports[`schema function body schema.only Should only validate included fields 1`] = `
Object {
  "a": Object {
    "errors": Array [
      "\\"a\\" must be a number",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
}
`;

exports[`schema function body schema.only Should only validate included fields 2`] = `
Object {
  "a": Array [
    "\\"a\\" must be a number",
  ],
}
`;

exports[`schema function body schema.skip Should validate all but skipped fields 1`] = `
Object {
  "b": Object {
    "errors": Array [
      "\\"b\\" must be a numeric string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c": Object {
    "errors": Array [
      "id must be a string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[0]": Object {
    "errors": Array [
      "id must be a string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[0].id": Object {
    "errors": Array [],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[1]": Object {
    "errors": Array [
      "id must be a string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[1].id": Object {
    "errors": Array [],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[2]": Object {
    "errors": Array [],
    "hasErrors": false,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[2].id": Object {
    "errors": Array [],
    "hasErrors": false,
    "hasWarnings": false,
    "warnings": Array [],
  },
}
`;

exports[`schema function body schema.skip Should validate all but skipped fields 2`] = `
Object {
  "b": Array [
    "\\"b\\" must be a numeric string",
  ],
  "c": Array [
    "id must be a string",
  ],
  "c[0]": Array [
    "id must be a string",
  ],
  "c[0].id": Array [],
  "c[1]": Array [
    "id must be a string",
  ],
  "c[1].id": Array [],
  "c[2]": Array [],
  "c[2].id": Array [],
}
`;

exports[`schema function body schema.skip Should validate all but skipped fields 3`] = `
Object {
  "a": Object {
    "errors": Array [
      "\\"a\\" must be a number",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "b": Object {
    "errors": Array [
      "\\"b\\" must be a numeric string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
}
`;

exports[`schema function body schema.skip Should validate all but skipped fields 4`] = `
Object {
  "a": Array [
    "\\"a\\" must be a number",
  ],
  "b": Array [
    "\\"b\\" must be a numeric string",
  ],
}
`;

exports[`schema function body schema.skip Should validate all but skipped fields 5`] = `
Object {
  "c": Object {
    "errors": Array [
      "id must be a string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[0]": Object {
    "errors": Array [
      "id must be a string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[0].id": Object {
    "errors": Array [],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[1]": Object {
    "errors": Array [
      "id must be a string",
    ],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[1].id": Object {
    "errors": Array [],
    "hasErrors": true,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[2]": Object {
    "errors": Array [],
    "hasErrors": false,
    "hasWarnings": false,
    "warnings": Array [],
  },
  "c[2].id": Object {
    "errors": Array [],
    "hasErrors": false,
    "hasWarnings": false,
    "warnings": Array [],
  },
}
`;

exports[`schema function body schema.skip Should validate all but skipped fields 6`] = `
Object {
  "c": Array [
    "id must be a string",
  ],
  "c[0]": Array [
    "id must be a string",
  ],
  "c[0].id": Array [],
  "c[1]": Array [
    "id must be a string",
  ],
  "c[1].id": Array [],
  "c[2]": Array [],
  "c[2].id": Array [],
}
`;
